research/delf/delf/python/training/global_features/train_utils.py

```diff
@@ -0,0 +1,368 @@
+        break
+
+      output = tf.zeros(shape=(0, model.meta['outputdim']), dtype=tf.float32)
+      for input_tuple in input:
+        for img in input_tuple:
+          # Compute the global descriptor vector.
+          model_out = model(tf.expand_dims(img, axis=0), training=False)
+          output = tf.concat([output, model_out], 0)
+
+      # No need to reduce memory consumption (no backward pass):
+      # Compute loss for the full batch.
+      tmp_target = tf.concat(target, axis=0)
+      queries = tf.boolean_mask(output, tmp_target == -1, axis=0)
+      positives = tf.boolean_mask(output, tmp_target == 1, axis=0)
+      negatives = tf.boolean_mask(output, tmp_target == 0, axis=0)
+      negatives = tf.reshape(negatives, [tf.shape(queries)[0], neg_num,
+                                         model.meta['outputdim']])
+      loss = criterion(queries, positives, negatives)
+
+      # Record loss.
+      losses.update(loss / batch_size, batch_size)
+
+    # Measure elapsed time.
+    batch_time.update(time.time() - end)
+    end = time.time()
+
+    # Record immediate loss and elapsed time.
+    if debug and ((batch_num + 1) % print_frequency == 0 or
+                  batch_num == 0 or (batch_num + 1) == all_batch_num):
+      global_features_utils.debug_and_log(
+              '>> {0}: [{1} epoch][{2}/{3} batch]\t Time val: {'
+              'batch_time.val:.3f} '
+              '(Batch Time avg: {batch_time.avg:.3f})\t Data {'
+              'data_time.val:.3f} ('
+              'Time avg: {data_time.avg:.3f})\t Immediate loss value: {'
+              'loss.val:.4f} '
+              '(Loss avg: {loss.avg:.4f})'.format(
+                      state, epoch, batch_num + 1, all_batch_num,
+                      batch_time=batch_time,
+                      data_time=data_time, loss=losses), debug=True, log=False)
+    batch_num += 1
+
+  return losses.avg
+
+
+def test(datasets, net, epoch, writer=None, model_directory=None,
+         precompute_whitening=None, data_root='data', multiscale=[1.]):
+  """Testing step.
+
+  Evaluates the network on the provided test datasets by computing single-scale
+  mAP for easy/medium/hard cases. If `writer` is specified, saves the mAP
+  values in a tensorboard supported format.
+
+  Args:
+    datasets: List of dataset names for model testing (from
+      `_TEST_DATASET_NAMES`).
+    net: Network to evaluate.
+    epoch: Integer, epoch number.
+    writer: Tensorboard writer.
+    model_directory: String, path to the model directory.
+    precompute_whitening: Dataset used to learn whitening. If no
+      precomputation required, then `None`.
+    data_root: Absolute path to the data folder.
+    multiscale: List of scales for multiscale testing.
+  """
+  global_features_utils.debug_and_log(">> Testing step:")
+  global_features_utils.debug_and_log(
+          '>> Evaluating network on test datasets...')
+
+  # For testing we use image size of max 1024.
+  image_size = 1024
```

Comment:
generally it's better to avoid such hard-codings. Can we maybe move this into an argument with default 1024?